compiler/backend/src/org/jetbrains/kotlin/codegen/DefaultParameterValueSubstitutor.kt

```diff
@@ -0,0 +1,204 @@
+        val context = owner.intoFunction(functionDescriptor)
+
+        for (i in 1..count) {
+            generateOverloadWithSubstitutedParameters(functionDescriptor, delegateFunctionDescriptor, classBuilder, function, context, i)
+        }
+    }
+
+    private fun FunctionDescriptor.countDefaultParameters() =
+        getValueParameters().count { it.hasDefaultValue() }
+
+    /**
+     * Generates an overload for [functionDescriptor] that substitutes default values for the last
+     * [substituteCount] parameters that have default values.
+     *
+     * @param functionDescriptor the method for which the overloads are generated
+     * @param delegateFunctionDescriptor the method descriptor for the implementation that we need to call
+     *     (same as [functionDescriptor] in all cases except for companion object methods annotated with [platformStatic],
+     *     where [functionDescriptor] is the static method in the main class and [delegateFunctionDescriptor] is the
+     *     implementation in the companion object class)
+     * @param methodElement the PSI element for the method implementation (used in diagnostic messages only)
+     */
+    fun generateOverloadWithSubstitutedParameters(functionDescriptor: FunctionDescriptor,
+                                                  delegateFunctionDescriptor: FunctionDescriptor,
+                                                  classBuilder: ClassBuilder,
+                                                  methodElement: JetElement?,
+                                                  context: CodegenContext<*>,
+                                                  substituteCount: Int) {
+        val isStatic = AsmUtil.isStaticMethod(context.getContextKind(), functionDescriptor)
+        val flags = AsmUtil.getVisibilityAccessFlag(functionDescriptor) or (if (isStatic) Opcodes.ACC_STATIC else 0)
+        val remainingParameters = getRemainingParameters(functionDescriptor.getOriginal(), substituteCount)
+        val signature = state.getTypeMapper().mapSignature(functionDescriptor, context.getContextKind(),
+                                                           remainingParameters)
+        val mv = classBuilder.newMethod(OtherOrigin(functionDescriptor), flags,
+                                        signature.getAsmMethod().getName(),
+                                        signature.getAsmMethod().getDescriptor(), null,
+                                        FunctionCodegen.getThrownExceptions(functionDescriptor, state.getTypeMapper()))
+
+        if (state.getClassBuilderMode() == ClassBuilderMode.LIGHT_CLASSES) return
+
+        val frameMap = FrameMap()
+        val v = InstructionAdapter(mv)
+        mv.visitCode()
+
+        val methodOwner = state.getTypeMapper().mapToCallableMethod(delegateFunctionDescriptor, false, context).getOwner()
+        if (!isStatic) {
+            frameMap.enterTemp(AsmTypes.OBJECT_TYPE)
+            v.load(0, methodOwner) // Load this on stack
```

Comment:
Maybe use the value returned by `enterTemp` instead of `0` for better readability